packages/server/lib/browsers/chrome.ts

```diff
@@ -329,10 +328,123 @@ const _handleDownloads = async function (client, dir, automation) {
   })
 }
 
-const _setAutomation = (client, automation) => {
-  return automation.use(
-    new CdpAutomation(client.send, client.on, automation),
-  )
+let frameTree
+let gettingFrameTree
+
+// eslint-disable-next-line @cypress/dev/arrow-body-multiline-braces
+const _updateFrameTree = (client) => async () => {
+  debug('update frame tree')
+
+  gettingFrameTree = new Promise<void>(async (resolve) => {
+    frameTree = (await client.send('Page.getFrameTree')).frameTree
+
+    debug('frame tree updated')
+
+    gettingFrameTree = null
+
+    resolve()
+  })
+}
+
+// we can't get the frame tree during the Fetch.requestPaused event, because
+// the CDP is tied up during that event and can't be utilized. so we maintain
+// a reference to it that's updated when it's likely to have been changed
+const _listenForFrameTreeChanges = (client) => {
+  debug('listen for frame tree changes')
+
+  client.on('Page.frameAttached', _updateFrameTree(client))
```

Comment:
@chrisbreiding this is like... the edgiest edge case ever, but this reminds me of the conversation we had about re-sending `*.enable` events - won't this come out of sync if a frame is attached/detached while the laptop is suspending...? might be worth noting in the source code in case it ever comes up, i don't think it's worth the effort to fix until we see bug reports, thoughts?

```suggestion
  // NOTE: an edge case can occur where these events are emitted when CDP is temporarily offline
  // (for example, a laptop waking from/going to sleep) and the frame tree gets out of sync
  client.on('Page.frameAttached', _updateFrameTree(client))
```

Reply:
i guess it would be really simple to fix for real, could just add a `client.on('reconnect')` event or similar

Reply:
Addressed in https://github.com/cypress-io/cypress/pull/21137/commits/02f7fd849c048708a2b123ae59f4a3595beb3002

Reply:
Very nice!